include/ppx/scene/scene_resource_manager.h

```diff
@@ -0,0 +1,126 @@
+// duplicate objects.
+//
+// The resource manager acts as the external owner of all shared resources
+// for scenes and meshes. Required objects can be shared in a variety of
+// difference cases:
+//   - images and image views can be shared among textures
+//   - textures can be shared among materials by way of texture views
+//   - materials can be shared among primitive batches
+//   - mesh data can be shared among meshes
+//   - meshes can be shared among nodes
+//
+// Both scene::Scene and scene::Mesh will call ResourceManager::DestroyAll()
+// in their destructors to destroy all its references to shared objects.
+// If afterwards a shared object has an external reference, the code holding
+// the reference is responsible for the shared objct.
+//
+class ResourceManager
+{
+public:
+    ResourceManager();
+    virtual ~ResourceManager();
+
+    bool Find(uint64_t objectId, scene::ImageRef& outObject) const;
+    bool Find(uint64_t objectId, scene::SamplerRef& outObject) const;
+    bool Find(uint64_t objectId, scene::TextureRef& outObject) const;
+    bool Find(uint64_t objectId, scene::MaterialRef& outObject) const;
+    bool Find(uint64_t objectId, scene::MeshDataRef& outObject) const;
+    bool Find(uint64_t objectId, scene::MeshRef& outObject) const;
+
+    // Cache functions assumes ownership of pObject
+    ppx::Result Cache(uint64_t objectId, const scene::ImageRef& object);
+    ppx::Result Cache(uint64_t objectId, const scene::SamplerRef& object);
+    ppx::Result Cache(uint64_t objectId, const scene::TextureRef& object);
+    ppx::Result Cache(uint64_t objectId, const scene::MaterialRef& object);
+    ppx::Result Cache(uint64_t objectId, const scene::MeshDataRef& object);
+    ppx::Result Cache(uint64_t objectId, const scene::MeshRef& object);
+
+    uint32_t GetImageCount() const { return static_cast<uint32_t>(mImages.size()); }
+    uint32_t GetSamplerCount() const { return static_cast<uint32_t>(mSamplers.size()); }
+    uint32_t GetTextureCount() const { return static_cast<uint32_t>(mTextures.size()); }
+    uint32_t GetMaterialCount() const { return static_cast<uint32_t>(mMaterials.size()); }
+    uint32_t GetMeshDataCount() const { return static_cast<uint32_t>(mMeshData.size()); }
+    uint32_t GetMeshCount() const { return static_cast<uint32_t>(mMeshes.size()); }
+
+    void DestroyAll();
+
+private:
+    template <
+        typename ObjectT,
+        typename ObjectRefT = std::shared_ptr<ObjectT>>
+    bool FindObject(
+        uint64_t                                        objectId,
+        const std::unordered_map<uint64_t, ObjectRefT>& container,
+        ObjectRefT&                                     outObject) const
+    {
+        auto it = container.find(objectId);
+        if (it == container.end()) {
+            return false;
+        }
+
+        outObject = (*it).second;
+        return true;
+    }
+
+    template <
+        typename ObjectT,
+        typename ObjectRefT = std::shared_ptr<ObjectT>>
```

Comment:
I'm not sure I follow what's happening here.
It seems this class should own the objects they store.
But we pass a reference to a `shared_ptr` when storing a value. Which strongly implies "I'm sharing ownership with you, not releasing to you".

But what's stored is a shared_ptr, meaning we do share ownership.
However, this doesn't handle allocation manually, which makes me wonder why the dtor is manually called in the cleanup function down.

Reply:
+1, is there any specific reason to use shared_ptr here rather than weak_ptr or raw pointer? since the resource manager is actually deleting the objects, rather than relying on shared_ptr

Reply:
What I really want here is a non-owning reference count but I couldn't find a C++ pattern for it other than maybe using std::weak_ptr. The reference count is useful if an application wants to sweep delete objects that aren't being used by other objects.

My reading on std::weak_ptr was that it doesn't affect the reference count unless it was converted to std::shared_ptr for usage. 

Open to other approaches though.

Reply:
dtor was left over from one of the refactors. It's been removed.

Reply:
I'd be fine with the weak_ptr, as it shows there is no ownership.

Reply:
weak_ptr doesn't do reference counting. It's not in there yet, but eventually the scene::Scene will have functions to allow  add/remove objects outside the context of scene loading. This is where the reference count lets utility functions know if an object is used at all, for example if an application wants to sweep remove all images that aren't referenced by any scene objects.

Obviously, this doesn't have to be shared_ptr, it's just the only facility available right now.

I've removed all external access to the reference objects so there should be no leaks. I have a different design in mind using a non-owning reference count. But we can refactor to that later. There's a few other projects needing the scene + GLTF loading sooner than later.